First merge

sachinhosmani opened the pull request:
`method-list` (for adding breakpoints) isn't ready yet because `core-selector`  is buggy.
The plan is to put `method-list` and `object-tree` on the right side in separate tabs. But `paper-tabs` was  not being compliant with `vulanize`, so that's not done yet.

Requires `O.o()` to work.

Thank you

<b>Update for #2:</b>
`core-selector` still needs to be rewritten and so `method-list` (for breakpoints) is kind of buggy.
Issue with `paper-tabs` got sorted out, so there are tabs on the right now.

sachinhosmani committed:
Initial commit

sachinhosmani committed:
Shows the DOM tree as a nested HTML list

sachinhosmani committed:
Added element-tree. Shows DOM tree when panel is opened and updates tree when page is refreshed.

sachinhosmani committed:
Some Polymer JSON serialization WIP

sachinhosmani committed:
Working element tree with all DOM nodes. UI shows selection.

sachinhosmani committed:
Side pane added. Object tree added. Shows properties in response to selections in element tree. Properties need better selection.

sachinhosmani committed:
Data binding basically work.

sachinhosmani committed:
added comments. fixed bugs in editable-label, added a 'type' field in the object inspector

sachinhosmani committed:
Composed DOM constructed. JSONizer broken. Element and Object trees - change in progress.

sachinhosmani committed:
...

sachinhosmani committed:
Attempt to fix object tree cyclical dependencies in progress

sachinhosmani committed:
Cyclical references taken care of. Object tree, Element tree look fine

sachinhosmani committed:
Attempt to use polymer-ui-splitter, doesn't work well

sachinhosmani committed:
Data-binding fixed. Many props found missing.

sachinhosmani committed:
uses Object.observe to keep object tree updated. fixed a few bugs.

sachinhosmani committed:
Rewrote object-tree using data-binding completely. couldn't test O.o(): no source build at home

sachinhosmani committed:
Fixed object-tree. Everything looks good in it

sachinhosmani committed:
WIP: Writing an eval-helper

sachinhosmani committed:
Fixed broken state after EvalHelper. Realize there is a bug with reflection + O.o()

sachinhosmani committed:
attempted to fix the bug with reflection. need to deal with getters/setters

sachinhosmani committed:
Added a refresh button to getters/setter. Fixed dumb bug. Found problems with unwanted properties being added in response to using setters.

sachinhosmani committed:
WIP: Adding a method-list

sachinhosmani committed:
Added method-list, but buggy. also have issues with paper-tab.

sachinhosmani committed:
Tried to use core-selector and paper-tabs but failed. templates now highlighted to show the model behind it (WIP). Breakpoints support added, but buggy because of core-selector

sachinhosmani committed:
Fixed removing O.o() listeners. Adds srcURLs to eval code. Fixed many other bugs. Fixed switching between selected elements.

sachinhosmani committed:
Fixed bugs with removing O.o() listeners

sachinhosmani committed:
Code tweaking

sachinhosmani committed:
...

sachinhosmani committed:
DOM mutations observed. A little bug found in

sachinhosmani committed:
Added colors to element-tree. Highlights elements when hovered. Fix in DOM mutation observer.

sachinhosmani committed:
Quick bug fix in

sachinhosmani committed:
Properties sorted correctly, fix in refresh property, Polymer and published properties shown separately, added color to element-tree

sachinhosmani committed:
A little code clean-up and comments

sachinhosmani committed:
Clean-up after closing pane. Code tweaks.

sachinhosmani committed:
removed bower_components/. unhighlights selected element when devtools closed

sachinhosmani committed:
Added readme

sachinhosmani committed:
Actually removed bower_components

sachinhosmani committed:
Update README.md

sachinhosmani committed:
Uses core-icon

sachinhosmani committed:
Merge with update remote (Readme update on Github)

alice commented on DOMSerializer.js:
```diff
@@ -0,0 +1,342 @@
+  }
+
+  // Polymer-specific stuff (to flag them differently)
+  var polymerSpecificProps = {
```
How was this list generated? How will it be kept up to date?

sachinhosmani replied:
It was written manually, copied from the docs. I think I'll have to manually keep it up-to-date.
I don't think it will change very often.

alice replied:
Perhaps comment with a link to the relevant docs, for any future maintainers.

alice commented on background.js:
```diff
@@ -0,0 +1,64 @@
+
+  // All the communcation
+  chrome.runtime.onMessage.addListener(function (message, sender, sendResponse) {
+    var port = tabIdToPortMap[sender.tab.id];
```
Since you're checking for port in each case below, you could just early-out here, e.g.
var port = ...
if (!port)
  return;
switch ...

alice commented on contentscript.js:
```diff
@@ -0,0 +1,28 @@
+  if (message.name === 'clean-up') {
+    window.dispatchEvent(new CustomEvent('clean-up'));
+  }
+});
```
Newline here :)

alice commented on manifest.json:
```diff
@@ -0,0 +1,22 @@
+{
+  "name" : "Test extension",
```
This name is a bit misleading! :)

alice commented on manifest.json:
```diff
@@ -0,0 +1,22 @@
+    "scripts": ["background.js"]
+  },
+  "devtools_page": "devtools.html",
+  "content_scripts" : [
```
You should inject content scripts programmatically instead of matching every page like this: https://developer.chrome.com/extensions/content_scripts#pi

mangini replied:
Yes, absolutely! The way it is now, every page has the content script injected, even if devtools is closed.

alice commented on manifest.json:
```diff
@@ -0,0 +1,22 @@
+      "run_at" : "document_idle"
+    }
+  ],
+  "permissions": ["http://*/*", "https://*/*"],
```
"&lt;all_urls&gt;" will get you access to all URLs.

alice commented on DOMSerializer.js:
```diff
@@ -0,0 +1,342 @@
+  * Checks if a property is an acessor. obj.hasOwnProperty(prop) has to be true.
+  */
+  function propHasAccessor (obj, prop) {
+     var descriptor = Object.getOwnPropertyDescriptor(obj, prop);
```
The formatting is a little off here, and some of your other formatting (e.g. the JSDoc comments) is a little non-standard. Have you considered using a formatter like jsfmt (https://github.com/rdio/jsfmt) which can be used as a plug-in for vim/emacs and/or run on the command line, for example as a git hook?

alice commented on panel-orig.js:
```diff
@@ -0,0 +1,402 @@
+            if (error) {
+              throw error;
+            }
+            EvalHelper.executeFunction('getDOMString', [], function (result, error) {
```
Did you intend to ignore |error| here?

sachinhosmani replied:
No I missed that.

alice commented on evalHelper.js:
```diff
@@ -0,0 +1,116 @@
+    delete window._polymerNamespace_;
+  }
+  // Wait till the namespace is created and clean-up handler is created.
+  chrome.devtools.inspectedWindow.eval('window._polymerNamespace_ = {};',
```
This looks dangerous to me. There's a non-zero chance that an author could use a variable with this name. I would suggest that if this absolutely needs to be set in the main world, you should use a GUID (such as the extension ID) to ensure that the name is unique.

sachinhosmani committed:
Selected element in inspector is the reflected in the pane. bower.json version fix. Added a persistent-button as an  in page (unsure about using it)

sachinhosmani committed:
Improved method-list UI

alice commented on DOMSerializer.js:
```diff
@@ -0,0 +1,342 @@
+        type: 'error',
+        hasAccessor: true,
+        error: true,
+        value: error.message,
```
s/error/e/ ?

sachinhosmani replied:
I'm sorry, what does that mean?

mangini replied:
regular expression. it's a way to say 'replace "error" by "e"'.

alice replied:
Right :) Sorry, I should have been more clear.
For historical context, it comes from sed: http://en.wikipedia.org/wiki/Sed#Substitution_command but it's also used in vi[m] and Perl (and probably other places too).

alice commented on DOMSerializer.js:
```diff
@@ -0,0 +1,342 @@
+  * Copies a property from oldObj to newObj and adds some metadata.
+  * protoObject is the exact object in the chain where the prop is present.
+  * It is necessary to determine if the prop is an accessor or not.
+  */
```
I'd suggest using standard Closure-style JSDoc type annotations for your params, here and everywhere else. It'll mean that if you want to use a type-checking builder somewhere down the line, you'll have the annotations all ready for all your methods, but also it provides useful documentation for understanding the code as well.

mangini replied:
Can you rename newObj to something that is not misleading? I spent some time trying to understand what is this push() method on Object :-)

sachinhosmani committed:
added a model-tree

sachinhosmani committed:
Added tabs to right-pane. Fixed minor bug in getComposedTree. Wrote a 'core-pages'.

sachinhosmani committed:
added a border around tabs

mangini commented on DOMSerializer.js:
```diff
@@ -0,0 +1,346 @@
+     if (!descriptor) {
+      console.error(prop);
+     }
+     return !!descriptor.set || !!descriptor.get;
```
!! is too obscure. Please, change it to something else, like
 return descriptor.set ? true : false || descriptor.get ? true : false;

alice replied:
I'd suggest Boolean(descriptor.set) || Boolean(descriptor.get) rather than the ternary. (Note that Boolean(foo) returns a primitive, not a new Boolean wrapper object.)

mangini replied:
+1, thanks Alice.

mangini commented on DOMSerializer.js:
```diff
@@ -0,0 +1,346 @@
+        while (proto && !Polymer.isBase(proto)) {
+          var props = getPolymerProps(proto);
+          props = props.filter(checkAdded);
+          props.forEach(addToAddedProps);
```
Any reason why you can't set addedProps[el] in the for loop below? The way it is, you are iterating over props twice, (apparently) for no reason. You can make it even better by checking if it is added (checkAdded) and continuing inside the "for" loop, instead of using the "props.filter" line above, reducing from three to one iteration over props.

mangini commented on DOMSerializer.js:
```diff
@@ -0,0 +1,346 @@
+          }
+        }
+      }
+      // copyProperty(Object.prototype, obj, destObj, '__proto__');
```
Get rid of commented code, or add TODO annotations.

mangini commented on DOMSerializer.js:
```diff
@@ -0,0 +1,346 @@
+    }
+
+    /**
+    * Explores an array for proerties
```
typo

mangini commented on DOMSerializer.js:
```diff
@@ -0,0 +1,346 @@
+        children.push(root.children[i]);
+      }
+    }
+    return children.filter(function (node) {
```
Can't you test before pushing to the array? filtering is not O(1) and requires both memory allocation and CPU. Since this method is called recursively, it seems to me that moving filter code to a in-loop check would also avoid redundant checks.

mangini commented on DOMSerializer.js:
```diff
@@ -0,0 +1,346 @@
+  *   isPolymer: <true|false>,
+  *   children: [<more such objects>]
+  * }
+  * @callback: is execuated on every DOM element found
```
nit: typo

mangini commented on DOMSerializer.js:
```diff
@@ -0,0 +1,346 @@
+  *   children: [<more such objects>]
+  * }
+  * @callback: is execuated on every DOM element found
+  * @isPolymer is supposed to tell if an element is a Polymer element
```
Strange syntax for commenting isPolymer. This seems like isPolymer is a param of this method, not a field in the returned object.

sachinhosmani replied:
Oh, isPolymer used to get passed in as a param an older revision. Forgot to update the comment.

mangini commented on README.md:
```diff
@@ -0,0 +1,31 @@
+Polymer devtools extension
```
Please, add a license file, like [this](https://github.com/GoogleChrome/chrome-app-samples/blob/master/LICENSE)

(add it as a separate file)

mangini commented on bower.json:
```diff
@@ -0,0 +1,18 @@
+{
+  "name": "test",
```
Set project name and change the license to "Apache-2.0".

mangini commented on evalHelper.js:
```diff
@@ -0,0 +1,116 @@
+  * Convert any object to a string
+  */
+  function serialize (object) {
+    function serializeArray (arr) {
```
Can't you use JSON.stringify instead?

mangini commented on hostPageHelpers.js:
```diff
@@ -0,0 +1,746 @@
+  }
+}
+
+/*function addElementSelectionCSS (key) {
```
Get rid of commented code or add TODOs

mangini commented on hostPageHelpers.js:
```diff
@@ -0,0 +1,746 @@
+function highlight (key, isHover) {
+  var element = window._polymerNamespace_.DOMCache[key];
+  if (isHover) {
+    window._polymerNamespace_.prevHoveredOutline = element.style.outline;
```
Set the namespace object in a variable, so you don't need to repeat this every time (a typo would be very hard to catch). Also, it makes it easier to implement Alice's suggestion to prefix with a GUID.

sachinhosmani replied:
Since all this code is injected after .toString(), the namespace has to be in the functions as they are. If I were to use a variable to hold the namespace name, then I'd have to make sure that variable is defined in the hose page too.

And what @alice suggested is a little difficult since I'd have to type in the GUID everywhere.
Maybe I could pass the namespace name (from the extension) to every host page function?
Or I could do a string replace on every function after .toString()ing it and before executing it?

alice replied:
Actually, I think you might want to put your functions into the `_polymerNamespace_` object as well, otherwise you're spamming the main world namespace with the function names as well. Recall that `function foo() {...}` will be placed on `window` if it's not executed inside a closure.

So with that in mind, you might want to take some inspiration from the IIFE (immediately-invoked function expression) structure, somewhat similar to what is used in Polymer itself, e.g. https://github.com/Polymer/polymer-dev/blob/master/src/api.js

So you could have something like:

```
var highlightFunction = function(namespace) {
  function highlight(key, isHover) {
    var element = namespace.DOMCache[key];
    ...
  }

  namespace.highlight = highlight;
};
```

and then when you inject it, you could wrap it in the IIFE structure like this (obviously not exactly like this, but you get the idea):

```
var namespace = _polymerNamespace_ + chrome.runtime.id;
var toEval = '(' + highlightFunction.toString() + ')(' + namespace + ')';
chrome.devtools.inspectedWindow.eval(toEval, ...);
```

sachinhosmani replied:
All those functions do go into `_polymerNamespace_`. EvalHelper.js does that.

And `IIFE` should do the trick. Thanks!

I think it would be inconvenient to wrap all functions like that. How about

```
'(function (namespace) {' + fnString + fnName + '();' +  ' })();'
```

do define a function ?

And all functions in `hostPageHelpers.js` will use `window[namespace].` and not `window._polymerNamespace`.

alice replied:
That doesn't look quite right - where are you passing in the namespace? And why does fnString come before fnName?

sachinhosmani replied:
I'm so sorry, that was rubbish.

```
var NAMESPACE = chrome.runtime.id;
'(function (namespace) { window["' + NAMESPACE + '"].' + fnName + ' = ' + fnString + ';})("' + NAMESPACE + '");'
```

mangini commented on hostPageHelpers.js:
```diff
@@ -0,0 +1,746 @@
+// TODO: Improve blacklist
+function setBlacklist () {
+  window._polymerNamespace_.filterProperty.blacklist = {
+  accessKey: true,
```
Can you describe what is this blacklist for, and where did you get these elements, so it is easier to maintain it in the future?

sachinhosmani replied:
All those properties are HTML generic properties (all HTMLElements have them). They're not supposed to be in this level of prototype chain, but they are (it is a webkit bug apparently).
We don't want to show them in the object tree.

Alright, I'll put a link to the docs.

mangini commented on hostPageHelpers.js:
```diff
@@ -0,0 +1,746 @@
+          domNode.__keyPolymer__ = key;
+          converted.key = key;
+
+          if (key === 1 || domNode.shadowRoot) {
```
Describe what means "key === 1"

mangini commented on object-tree/object-tree.html:
```diff
@@ -0,0 +1,85 @@
+        margin-left: 4px;
+      }
+    </style>
+    <!--IDs are assigned dynamically and uniquely. They serve to identify the element.-->
```
I didn't understand why a static id (id='content') is set right after this comment. It seems that either the comment is wrong or the property is unnecessarily set.

sachinhosmani replied:
There is CSS applying to that ID.

The comment applies to the dynamically assigned IDs. You're right, it seems odd to find a static ID right after that. I guess I'll just move the comment one line down.

mangini commented on panel-orig.js:
```diff
@@ -0,0 +1,444 @@
+  function expandObject (path, isModel) {
+    var key = elementTree.selectedChild.key;
+    EvalHelper.executeFunction('getObjectString', [key, path, isModel], function (result, error) {
+      var props = JSON.parse(result.data).value;
```
if (error) { throw error } ?

mangini commented on panel-orig.js:
```diff
@@ -0,0 +1,444 @@
+      var index = event.detail.index;
+      var functionName = event.detail.isSet ? 'setBreakpoint' : 'clearBreakpoint';
+      EvalHelper.executeFunction(functionName, [key, [index]], function (result, error) {
+        if (error) {
```
Apparently all uses of EvalHelper.executeFunction rethrow error if it is not null. If that's the case, can't you add this behavior to EvalHelper? If there is an exception, you can always try/catch.

sachinhosmani replied:
At this stage shall leave them like this? I'll be able to quickly detect bugs this way. I wouldn't want anything silently failing right now.

mangini commented on panel-orig.js:
```diff
@@ -0,0 +1,444 @@
+            var childTree = isModel ? modelTree.tree : objectTree.tree;
+            try {
+              // If the observer reports changes before child-tree is ready, we must
+              // only
```
finish the comment :-)

mangini commented on panel.js:
```diff
@@ -0,0 +1,1361 @@
+
+    Polymer('core-selection', {
```
Is this a concatenation of Polymer core/paper elements? Can you describe this either in README or in this file? If possible, also mention how to update them (eg, concatenate files X, Y and Z from repo W)

sachinhosmani replied:
yes, it's created by vulcanize. I'm going to separate this into a build process, so it won't be necessary to have this file in the repo.

sachinhosmani committed:
Rewrote serializeDOM to support for dual DOM view

sachinhosmani committed:
Changed serialization for dual view. Added 'magnify button'. Shows unrendered elements.

sachinhosmani committed:
- Added dual view (local DOM + composed DOM)
- Got rid of unnecessary serializing
- Rewrote DOM fetching code
- Minor bug fixes
- (TODO: Need to make non-Polymer elements clickable to show their models)
- Issue found with seeing stuff inside `<shadow>` inside Local DOM

sachinhosmani committed:
Quick typo bug fix

sachinhosmani committed:
Tree scrolling fixes in.

sachinhosmani committed:
Added bread-crumbs

sachinhosmani committed:
Conditionally inject content scripts

sachinhosmani committed:
Added margin around bread-crumbs

sachinhosmani committed:
Added doc links

sachinhosmani committed:
Code formatting in response to review

sachinhosmani committed:
Typo fixes

sachinhosmani committed:
Change use JSON.stringify() instead of serialize in evalHelper

sachinhosmani committed:
minor tweaks and comment fixes

sachinhosmani committed:
Fixed bower.json

sachinhosmani committed:
Added LICENSE file

sachinhosmani committed:
Element tree expand button state change fix

sachinhosmani committed:
Uses extension ID as namespace

sachinhosmani committed:
Added namespaces to event names so it doesn't clash with any potential DOM events

sachinhosmani committed:
Added build system. It now uses Closure compiler.

sachinhosmani committed:
Updated README with new features and build instructions.

sachinhosmani committed:
Build copies core-splitter/handle.svg to build/ because vulcanize doesn't deal with resources

sachinhosmani committed:
Changed GitHub repo name. Updated README.

sachinhosmani committed:
Set allowOverflow='true' on core-splitter

addyosmani committed:
Strip unnecessary console log.

sachinhosmani committed:
Merge pull request #21 from addyosmani/master

Strip unnecessary console log.

sachinhosmani committed:
Extension works through page refresh, page change. Fixed tiny DOM mutations typo : 'changedElementKeys'.

sachinhosmani committed:
Merge branch 'master' with Addy's console.log strip PR.

sachinhosmani committed:
Update breadcrumbs before element-tree

sachinhosmani committed:
Added 'View definition file' feature for Polymer elements

sachinhosmani committed:
Added loading bars to tell if an element-tree is being loaded

sachinhosmani committed:
Show 'view source' only on hover

sachinhosmani committed:
Keeps track of DOM mutations while tree gets rendered, i.e., pending DOM mutations

sachinhosmani committed:
Added JSDoc comments

sachinhosmani committed:
Added more JSDoc comments

sachinhosmani committed:
Added more JSDoc comments

sachinhosmani committed:
Removed persistent-button

sachinhosmani committed:
Removed element test files

sachinhosmani committed:
More JSDocs and formatting

sachinhosmani committed:
Keeps silent in response to inspector element selection changes when in local DOM mode

sachinhosmani committed:
Comment fixes

sachinhosmani committed:
ReadMe update: View source button

sachinhosmani committed:
Newlines after .js files

sachinhosmani committed:
Element-tree: set 'polymer' attribute on '#name' with data-binding and not JS

sachinhosmani committed:
Added perf button. Functionally on the way.

alice commented on DOMJSONizer.js:
```diff
@@ -0,0 +1,442 @@
+
+// TODO: Remove all logic that pertains to Polymer elements from here and pass them as callbacks
+
+function DOMJSONizer () {
```
Nit: this style of a space between the method name and the parentheses is unusual. Since it's consistent, though, don't worry too much about fixing it unless it's straightforward to do so.

alice commented on DOMJSONizer.js:
```diff
@@ -0,0 +1,442 @@
+   * Checks if a property is an accessor property.
+   * @param  {Object} obj  The exact object on which the property is present.
+   * @param  {String} prop Name of the property
+   * @return {Boolean}     Whether the property is an accessor (get/set) or not.
```
nit: this is actually a small-b boolean - Boolean(foo) returns a boolean primitive.
(Other than that, though, great job on the JSDoc; makes it so much easier to see what's going on!)

alice commented on DOMJSONizer.js:
```diff
@@ -0,0 +1,442 @@
+  function propHasAccessor (obj, prop) {
+    var descriptor = Object.getOwnPropertyDescriptor(obj, prop);
+    if (!descriptor) {
+    console.error(prop);
```
nit: this line should be indented, and you should probably return something (false?) here.

alice commented on DOMJSONizer.js:
```diff
@@ -0,0 +1,442 @@
+  /**
+   * Checks if a property is an accessor property.
+   * @param  {Object} obj  The exact object on which the property is present.
+   * @param  {String} prop Name of the property
```
Similarly, {string} is actually a primitive, so this should be lower-case. https://developers.google.com/closure/compiler/docs/js-for-compiler#types

alice commented on DOMJSONizer.js:
```diff
@@ -0,0 +1,442 @@
+   * Copies a property from oldObj to newObjArray and adds some metadata.
+   * @param  {Object} protoObject  The exact object in chain where the property exists.
+   * @param  {Object} oldObj       The source object
+   * @param  {Array}  newObjArray  The destination object (which is maintained as an Array).
```
This same thing is referred to elsewhere as `destObj`. Probably best to be consistent. Ideally, something more descriptive  - see comment on line 202.

alice commented on DOMJSONizer.js:
```diff
@@ -0,0 +1,442 @@
+  }
+
+  /**
+   * Converts an object to JSON but only one-level deep.
```
Return type?

alice commented on DOMJSONizer.js:
```diff
@@ -0,0 +1,442 @@
+    // The root object is named as `Root`.
+    var res = {
+      name: 'Root',
+      value: []
```
I'd suggest calling this 'properties' rather than 'value'. Then, instead of references to `destObj` in the exploreFoo methods, you could refer to to `properties`, `objProperties` or similar.

alice commented on DOMJSONizer.js:
```diff
@@ -0,0 +1,442 @@
+          }
+        }
+      }
+      // TODO: `__proto__` ?
```
Could you expand more on this TODO? :)

sachinhosmani replied:
The object tree shows all the _own_ properties of a selected non-Polymer element (Polymer elements have properties explored across multiple levels of the prototype chain)
So there is no way someone can see properties in higher prototype objects. A good way to allow that is to show the `__proto__` property and allow the user to expand that.
It was an edge case here, since `__proto__` is not an own property of an element, but a getter defined on `Object.prototype`.

alice commented on DOMJSONizer.js:
```diff
@@ -0,0 +1,442 @@
+      function isAdded (el) {
+        return (el in addedProps);
+      }
+      function addToAddedProps (el) {
```
These two probably don't need to be functions.

sachinhosmani committed:
Added perf feature but only for . Fixed position of reload button. Fixed reload issues.

sachinhosmani committed:
Stylize console.log perf messages

sachinhosmani committed:
JSBeatified DOMJSONizer.js. Added : to console.log string in perf content script

alice commented on background.js:
```diff
@@ -0,0 +1,142 @@
+
+  // A panel just tried to connect to the extension background
+  chrome.runtime.onConnect.addListener(function(port) {
+    var portId;
```
Given tabId:portId is a 1:1 mapping, why not store tabId here instead (since it's available when you're initializing portId) and get rid of portId altogether?

alice commented on manifest.json:
```diff
@@ -0,0 +1,19 @@
+  "permissions": ["<all_urls>", "webNavigation"],
+  "content_security_policy": "script-src 'self' 'unsafe-eval'; object-src 'self'",
+  "manifest_version": 2,
+  "content_scripts": [
```
As commented previously: ideally, you should inject content scripts programmatically rather than injecting into all pages like this.

sachinhosmani replied:
I fixed that for the main content script.
But this one has to start as soon as it can. It needs to be there before `polymer-ready` fires.
I couldn't find a way to programmatically insert it sooner than `document_start` or at the same time as `document_start` (I couldn't find a way to detect a page reload that early).

alice replied:
Ahh, got it. Makes sense.

alice commented on DOMJSONizer.js:
```diff
@@ -0,0 +1,442 @@
+      // <content> must get replaced by what gets distributed into it
+      var children = root.getDistributedNodes();
+      for (var j = 0; j < children.length; j++) {
+        if ('tagName' in children[j]) {
```
Why are you checking for 'tagName'? Is this to see whether it's an element? I'd suggest using node.nodeType == Node.ELEMENT_NODE if that's the case.

alice commented on DOMJSONizer.js:
```diff
@@ -0,0 +1,442 @@
+    callback && callback(root, res, true);
+    if (root.tagName === 'CONTENT') {
+      // <content> must get replaced by what gets distributed into it
+      var children = root.getDistributedNodes();
```
suggest s/children/distributedNodes/ for clarity.

sachinhosmani committed:
Load trees asynchronously

sachinhosmani committed:
Set isTreeLoading correctly inside doTreeLoad

sachinhosmani committed:
Fixed async task ordering, pending actions. Typo in perfContentScript.js. Removed repetitive parts of 'init', moved it to 'createPageView' in panel-orig.js

sachinhosmani committed:
Store parent keys in each element in JSONized object. Show bread-crumbs for parent elements in local DOM view.

sachinhosmani committed:
Zoom out of local DOM view to last unchanged element when DOM mutation happens and local DOM root is affected.
Some concurrency updates with dom-mutation and page refresh.

paulirish committed:
add vulcanize to install directions.

sachinhosmani committed:
Merge pull request #31 from paulirish/patch-1

add vulcanize to install directions.

sachinhosmani committed:
Merge with Paul's PR.

oskbor committed:
Copy perfContentScript.js to build folder

Closes #32

The extension fails to install in Chrome without perfContentScript.js in the build folder.

Commit:
Add perfContentScript.js to the build

Commit:
Curse you Finder

addyosmani committed:
Merge pull request #34 from oskbor/patch-1

Copy perfContentScript.js to build folder

Commit:
Some hints for adding new functions to hostPageHelpers

Commit:
prefer closure-compiled perfContentScript

Commit:
Tinted overlay instead of an outline for highlighted elements.

Fixes #25.

Commit:
PR Feedback

nevir committed:
Merge pull request #35 from nevir/highlight

Tinted overlay instead of an outline for highlighted elements

Commit:
Don't scroll into view on highlight (but still do for selection)

sachinhosmani committed:
Scroll element into view upon selection

sachinhosmani committed:
Warn about unregistered custom elements

sachinhosmani committed:
Inline warningStyle

sachinhosmani committed:
Merge pull request #45 from sachinhosmani/master

Warn about unregistered custom elements

indolering committed:
Fixed Closure Compiler download link.

Closure Compiler download link was ~3 years out of date.

paulirish committed:
Merge pull request #53 from indolering/patch-1

Fixed Closure Compiler download link.
